{
 "metadata": {
  "name": "",
  "signature": "sha256:af3412f02e5a65382e9a3aea42d86668c1224e64672a015e92b4605d4fab4a61"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "# Chapter 1  Introduction"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "## 1.1  What is Biopython?"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "The Biopython Project is an international association of developers of freely available Python (http://www.python.org) tools for computational molecular biology. Python is an object oriented, interpreted, flexible language that is becoming increasingly popular for scientific computing. Python is easy to learn, has a very clear syntax and can easily be extended with modules written in C, C++ or FORTRAN.\n",
      "\n",
      "The Biopython web site (http://www.biopython.org) provides an online resource for modules, scripts, and web links for developers of Python-based software for bioinformatics use and research. Basically, the goal of Biopython is to make it as easy as possible to use Python for bioinformatics by creating high-quality, reusable modules and classes. Biopython features include parsers for various Bioinformatics file formats (BLAST, Clustalw, FASTA, Genbank,...), access to online services (NCBI, Expasy,...), interfaces to common and not-so-common programs (Clustalw, DSSP, MSMS...), a standard sequence class, various clustering modules, a KD tree data structure etc. and even documentation.\n",
      "\n",
      "Basically, we just like to program in Python and want to make it as easy as possible to use Python for bioinformatics by creating high-quality, reusable modules and scripts."
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "## 1.2  What can I find in the Biopython package"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "The main Biopython releases have lots of functionality, including:\n",
      "\n",
      "- The ability to parse bioinformatics files into Python utilizable data structures, including support for the following formats:\n",
      "    - Blast output \u2013 both from standalone and WWW Blast\n",
      "    - Clustalw\n",
      "    - FASTA\n",
      "    - GenBank\n",
      "    - PubMed and Medline\n",
      "    - ExPASy files, like Enzyme and Prosite\n",
      "    - SCOP, including \u2018dom\u2019 and \u2018lin\u2019 files\n",
      "    - UniGene\n",
      "    - SwissProt\n",
      "- Files in the supported formats can be iterated over record by record or indexed and accessed via a Dictionary interface.\n",
      "- Code to deal with popular on-line bioinformatics destinations such as:\n",
      "    - NCBI \u2013 Blast, Entrez and PubMed services\n",
      "    - ExPASy \u2013 Swiss-Prot and Prosite entries, as well as Prosite searches\n",
      "- Interfaces to common bioinformatics programs such as:\n",
      "    - Standalone Blast from NCBI\n",
      "    - Clustalw alignment program\n",
      "    - EMBOSS command line tools\n",
      "- A standard sequence class that deals with sequences, ids on sequences, and sequence features.\n",
      "- Tools for performing common operations on sequences, such as translation, transcription and weight calculations.\n",
      "- Code to perform classification of data using k Nearest Neighbors, Naive Bayes or Support Vector Machines.\n",
      "- Code for dealing with alignments, including a standard way to create and deal with substitution matrices.\n",
      "- Code making it easy to split up parallelizable tasks into separate processes.\n",
      "- GUI-based programs to do basic sequence manipulations, translations, BLASTing, etc.\n",
      "- Extensive documentation and help with using the modules, including this file, on-line wiki documentation, the web site, and the mailing list.\n",
      "- Integration with BioSQL, a sequence database schema also supported by the BioPerl and BioJava projects.\n",
      "\n",
      "We hope this gives you plenty of reasons to download and start using Biopython!"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "## 1.3  Installing Biopython"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "All of the installation information for Biopython was separated from this document to make it easier to keep updated.\n",
      "\n",
      "The short version is go to our downloads page (http://biopython.org/wiki/Download), download and install the listed dependencies, then download and install Biopython. Biopython runs on many platforms (Windows, Mac, and on the various flavors of Linux and Unix). For Windows we provide pre-compiled click-and-run installers, while for Unix and other operating systems you must install from source as described in the included README file. This is usually as simple as the standard commands:\n",
      "\n",
      "python setup.py build\n",
      "python setup.py test\n",
      "sudo python setup.py install\n",
      "(You can in fact skip the build and test, and go straight to the install \u2013 but its better to make sure everything seems to be working.)\n",
      "\n",
      "The longer version of our installation instructions covers installation of Python, Biopython dependencies and Biopython itself. It is available in PDF (http://biopython.org/DIST/docs/install/Installation.pdf) and HTML formats (http://biopython.org/DIST/docs/install/Installation.html)."
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "## 1.4  Frequently Asked Questions (FAQ)"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "1. How do I cite Biopython in a scientific publication? \n",
      "\n",
      "    Please cite our application note [1, Cock et al., 2009] as the main Biopython reference. In addition, please cite any publications from the following list if appropriate, in particular as a reference for specific modules within Biopython (more information can be found on our website):\n",
      "    - For the official project announcement: [13, Chapman and Chang, 2000];\n",
      "    - For Bio.PDB: [18, Hamelryck and Manderick, 2003];\n",
      "    - For Bio.Cluster: [14, De Hoon et al., 2004];\n",
      "    - For Bio.Graphics.GenomeDiagram: [2, Pritchard et al., 2006];\n",
      "    - For Bio.Phylo and Bio.Phylo.PAML: [9, Talevich et al., 2012];\n",
      "    - For the FASTQ file format as supported in Biopython, BioPerl, BioRuby, BioJava, and EMBOSS: [7, Cock et al., 2010].\n",
      "2. How should I capitalize \u201cBiopython\u201d? Is \u201cBioPython\u201d OK? \n",
      "\n",
      "    The correct capitalization is \u201cBiopython\u201d, not \u201cBioPython\u201d (even though that would have matched BioPerl, BioJava and BioRuby).\n",
      "3. What is going wrong with my print commands? \n",
      "\n",
      "    This tutorial now uses the Python 3 style print function. As of Biopython 1.62, we support both Python 2 and Python 3. The most obvious language difference is the print statement in Python 2 became a print function in Python 3.\n",
      "    \n",
      "    For example, this will only work under Python 2:\n",
      "    ```\n",
      "    >>> print \"Hello World!\"\n",
      "    Hello World!\n",
      "    ```\n",
      "    \n",
      "    If you try that on Python 3 you\u2019ll get a SyntaxError. Under Python 3 you must write:\n",
      "    ```\n",
      "    >>> print(\"Hello World!\")\n",
      "    Hello World!\n",
      "    ```\n",
      "\n",
      "    Surprisingly that will also work on Python 2 \u2013 but only for simple examples printing one thing. In general you need to add this magic line to the start of your Python scripts to use the print function under Python 2.6 and 2.7:\n",
      "    ```\n",
      "    from __future__ import print_function\n",
      "    ```\n",
      "    \n",
      "    If you forget to add this magic import, under Python 2 you\u2019ll see extra brackets produced by trying to use the print function when Python 2 is interpretting it as a print statement and a tuple.\n",
      "\n",
      "4. How do I find out what version of Biopython I have installed? \n",
      "\n",
      "    Use this:\n",
      "    ```\n",
      "    >>> import Bio\n",
      "    >>> print(Bio.__version__)\n",
      "    ...\n",
      "    ```\n",
      "  \n",
      "    If the \u201cimport Bio\u201d line fails, Biopython is not installed. If the second line fails, your version is very out of date. If the version string ends with a plus, you don\u2019t have an official release, but a snapshot of the in development code.\n",
      "5. Where is the latest version of this document?\n",
      "\n",
      "    If you download a Biopython source code archive, it will include the relevant version in both HTML and PDF formats. The latest published version of this document (updated at each release) is online:\n",
      "    - http://biopython.org/DIST/docs/tutorial/Tutorial.html\n",
      "    - http://biopython.org/DIST/docs/tutorial/Tutorial.pdf\n",
      "    \n",
      "    If you are using the very latest unreleased code from our repository you can find copies of the in-progress tutorial here:\n",
      "    - http://biopython.org/DIST/docs/tutorial/Tutorial-dev.html\n",
      "    - http://biopython.org/DIST/docs/tutorial/Tutorial-dev.pdf\n",
      "6. Why is the Seq object missing the upper & lower methods described in this Tutorial? \n",
      "\n",
      "    You need Biopython 1.53 or later. Alternatively, use str(my_seq).upper() to get an upper case string. If you need a Seq object, try Seq(str(my_seq).upper()) but be careful about blindly re-using the same alphabet.\n",
      "7. Why doesn\u2019t the Seq object translation method support the cds option described in this Tutorial? \n",
      "\n",
      "    You need Biopython 1.51 or later.\n",
      "8. What file formats do Bio.SeqIO and Bio.AlignIO read and write? \n",
      "\n",
      "    Check the built in docstrings (from Bio import SeqIO, then help(SeqIO)), or see http://biopython.org/wiki/SeqIO and http://biopython.org/wiki/AlignIO on the wiki for the latest listing.\n",
      "9. Why won\u2019t the Bio.SeqIO and Bio.AlignIO functions parse, read and write take filenames? They insist on handles! \n",
      "\n",
      "    You need Biopython 1.54 or later, or just use handles explicitly (see Section 22.1). It is especially important to remember to close output handles explicitly after writing your data.\n",
      "10. Why won\u2019t the Bio.SeqIO.write() and Bio.AlignIO.write() functions accept a single record or alignment? They insist on a list or iterator! \n",
      "\n",
      "    You need Biopython 1.54 or later, or just wrap the item with [...] to create a list of one element.\n",
      "11. Why doesn\u2019t str(...) give me the full sequence of a Seq object? \n",
      "\n",
      "    You need Biopython 1.45 or later. Alternatively, rather than str(my_seq), use my_seq.tostring() (which will also work on recent versions of Biopython).\n",
      "12. Why doesn\u2019t Bio.Blast work with the latest plain text NCBI blast output? \n",
      "\n",
      "    The NCBI keep tweaking the plain text output from the BLAST tools, and keeping our parser up to date is/was an ongoing struggle. If you aren\u2019t using the latest version of Biopython, you could try upgrading. However, we (and the NCBI) recommend you use the XML output instead, which is designed to be read by a computer program.\n",
      "13. Why doesn\u2019t Bio.Entrez.parse() work? The module imports fine but there is no parse function! \n",
      "\n",
      "    You need Biopython 1.52 or later.\n",
      "14. Why has my script using Bio.Entrez.efetch() stopped working? \n",
      "\n",
      "    This could be due to NCBI changes in February 2012 introducing EFetch 2.0. First, they changed the default return modes - you probably want to add retmode=\"text\" to your call. Second, they are now stricter about how to provide a list of IDs \u2013 Biopython 1.59 onwards turns a list into a comma separated string automatically.\n",
      "15. Why doesn\u2019t Bio.Blast.NCBIWWW.qblast() give the same results as the NCBI BLAST website? \n",
      "\n",
      "    You need to specify the same options \u2013 the NCBI often adjust the default settings on the website, and they do not match the QBLAST defaults anymore. Check things like the gap penalties and expectation threshold.\n",
      "16. Why doesn\u2019t Bio.Blast.NCBIXML.read() work? The module imports but there is no read function! \n",
      "\n",
      "    You need Biopython 1.50 or later. Or, use next(Bio.Blast.NCBIXML.parse(...)) instead.\n",
      "17. Why doesn\u2019t my SeqRecord object have a letter_annotations attribute? \n",
      "    \n",
      "    Per-letter-annotation support was added in Biopython 1.50.\n",
      "18. Why can\u2019t I slice my SeqRecord to get a sub-record? \n",
      "    \n",
      "    You need Biopython 1.50 or later.\n",
      "19. Why can\u2019t I add SeqRecord objects together? \n",
      "    \n",
      "    You need Biopython 1.53 or later.\n",
      "20. Why doesn\u2019t Bio.SeqIO.convert() or Bio.AlignIO.convert() work? The modules import fine but there is no convert function! \n",
      "    \n",
      "    You need Biopython 1.52 or later. Alternatively, combine the parse and write functions as described in this tutorial (see Sections 5.5.2 and 6.2.1).\n",
      "21. Why doesn\u2019t Bio.SeqIO.index() work? The module imports fine but there is no index function!\n",
      "\n",
      "    You need Biopython 1.52 or later.\n",
      "22. Why doesn\u2019t Bio.SeqIO.index_db() work? The module imports fine but there is no index_db function! \n",
      "\n",
      "    You need Biopython 1.57 or later (and a Python with SQLite3 support).\n",
      "23. Where is the MultipleSeqAlignment object? The Bio.Align module imports fine but this class isn\u2019t there! \n",
      "\n",
      "    You need Biopython 1.54 or later. Alternatively, the older Bio.Align.Generic.Alignment class supports some of its functionality, but using this is now discouraged.\n",
      "24. Why can\u2019t I run command line tools directly from the application wrappers? \n",
      "    \n",
      "    You need Biopython 1.55 or later. Alternatively, use the Python subprocess module directly.\n",
      "25. I looked in a directory for code, but I couldn\u2019t find the code that does something. Where\u2019s it hidden? \n",
      "\n",
      "    One thing to know is that we put code in __init__.py files. If you are not used to looking for code in this file this can be confusing. The reason we do this is to make the imports easier for users. For instance, instead of having to do a \u201crepetitive\u201d import like from Bio.GenBank import GenBank, you can just use from Bio import GenBank.\n",
      "26. Why does the code from CVS seem out of date? \n",
      "\n",
      "    In late September 2009, just after the release of Biopython 1.52, we switched from using CVS to git, a distributed version control system. The old CVS server will remain available as a static and read only backup, but if you want to grab the latest code, you\u2019ll need to use git instead. See our website for more details.\n",
      "    \n",
      "For more general questions, the Python FAQ pages http://www.python.org/doc/faq/ may be useful."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%HTML\n",
      "<style>\n",
      ".cell {\n",
      "    background-color: rgba(255,200,255,0.72);\n",
      "    font-family: Georgia, Cursive;\n",
      "    font-size: 144%;\n",
      "    border-radius: 10px;\n",
      "    spellcheck: true;\n",
      "}\n",
      "\n",
      "body {\n",
      "    background-color: #AAFFFF;\n",
      "    background-image: url('https://raw.github.com/chenleo/Resource/master/Images/FirstKiss1600X1200.jpg');\n",
      "    background-size: cover;\n",
      "    background-repeat: repeat;\n",
      "    background--attachment: fixed;\n",
      "}\n",
      "\n",
      "h1, h2 {\n",
      "    text-align: center;\n",
      "}\n",
      "\n",
      ".todo {\n",
      "    background-color: rgba(255,0,0,0.5);\n",
      "}\n",
      "</style>"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "<style>\n",
        ".cell {\n",
        "    background-color: rgba(255,200,255,0.72);\n",
        "    font-family: Georgia, Cursive;\n",
        "    font-size: 144%;\n",
        "    border-radius: 10px;\n",
        "    spellcheck: true;\n",
        "}\n",
        "\n",
        "body {\n",
        "    background-color: #AAFFFF;\n",
        "    background-image: url('https://raw.github.com/chenleo/Resource/master/Images/FirstKiss1600X1200.jpg');\n",
        "    background-size: cover;\n",
        "    background-repeat: repeat;\n",
        "    background--attachment: fixed;\n",
        "}\n",
        "\n",
        "h1, h2 {\n",
        "    text-align: center;\n",
        "}\n",
        "\n",
        ".todo {\n",
        "    background-color: rgba(255,0,0,0.5);\n",
        "}\n",
        "</style>"
       ],
       "metadata": {},
       "output_type": "display_data",
       "text": [
        "<IPython.core.display.HTML at 0x41a8eb8>"
       ]
      }
     ],
     "prompt_number": 1
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}